// @ezuikit/utils-i18n/dist/index.umd.js
!function(global, factory) {
    "object" == typeof exports && "undefined" != typeof module ? module.exports = factory() : "function" == typeof define && define.amd ? define(factory) : (global = "undefined" != typeof globalThis ? globalThis : global || self).I18n = factory();
}(this, function() {
    "use strict";
    function getDefaultExportFromCjs(x) {
        return x && x.__esModule && Object.prototype.hasOwnProperty.call(x, "default") ? x.default : x;
    }
    var cjs, hasRequiredCjs;
    var cjsExports = function() {
        if (hasRequiredCjs) return cjs;
        hasRequiredCjs = 1;
        var isMergeableObject = function(value) {
            return function(value) {
                return !!value && "object" == typeof value;
            }(value) && !function(value) {
                var stringValue = Object.prototype.toString.call(value);
                return "[object RegExp]" === stringValue || "[object Date]" === stringValue || function(value) {
                    return value.$$typeof === REACT_ELEMENT_TYPE;
                }(value);
            }(value);
        }, REACT_ELEMENT_TYPE = "function" == typeof Symbol && Symbol.for ? Symbol.for("react.element") : 60103;
        function cloneUnlessOtherwiseSpecified(value, options) {
            return !1 !== options.clone && options.isMergeableObject(value) ? deepmerge((val = value, 
            Array.isArray(val) ? [] : {}), value, options) : value;
            var val;
        }
        function defaultArrayMerge(target, source, options) {
            return target.concat(source).map(function(element) {
                return cloneUnlessOtherwiseSpecified(element, options);
            });
        }
        function getKeys(target) {
            return Object.keys(target).concat(function(target) {
                return Object.getOwnPropertySymbols ? Object.getOwnPropertySymbols(target).filter(function(symbol) {
                    return Object.propertyIsEnumerable.call(target, symbol);
                }) : [];
            }(target));
        }
        function propertyIsOnObject(object, property) {
            try {
                return property in object;
            } catch (_) {
                return !1;
            }
        }
        function mergeObject(target, source, options) {
            var destination = {};
            return options.isMergeableObject(target) && getKeys(target).forEach(function(key) {
                destination[key] = cloneUnlessOtherwiseSpecified(target[key], options);
            }), getKeys(source).forEach(function(key) {
                (function(target, key) {
                    return propertyIsOnObject(target, key) && !(Object.hasOwnProperty.call(target, key) && Object.propertyIsEnumerable.call(target, key));
                })(target, key) || (propertyIsOnObject(target, key) && options.isMergeableObject(source[key]) ? destination[key] = function(key, options) {
                    if (!options.customMerge) return deepmerge;
                    var customMerge = options.customMerge(key);
                    return "function" == typeof customMerge ? customMerge : deepmerge;
                }(key, options)(target[key], source[key], options) : destination[key] = cloneUnlessOtherwiseSpecified(source[key], options));
            }), destination;
        }
        function deepmerge(target, source, options) {
            (options = options || {}).arrayMerge = options.arrayMerge || defaultArrayMerge, 
            options.isMergeableObject = options.isMergeableObject || isMergeableObject, options.cloneUnlessOtherwiseSpecified = cloneUnlessOtherwiseSpecified;
            var sourceIsArray = Array.isArray(source);
            return sourceIsArray === Array.isArray(target) ? sourceIsArray ? options.arrayMerge(target, source, options) : mergeObject(target, source, options) : cloneUnlessOtherwiseSpecified(source, options);
        }
        return deepmerge.all = function(array, options) {
            if (!Array.isArray(array)) throw new Error("first argument should be an array");
            return array.reduce(function(prev, next) {
                return deepmerge(prev, next, options);
            }, {});
        }, cjs = deepmerge;
    }(), deepmerge = getDefaultExportFromCjs(cjsExports);
    function _create_class(Constructor, protoProps, staticProps) {
        return protoProps && function(target, props) {
            for (var i = 0; i < props.length; i++) {
                var descriptor = props[i];
                descriptor.enumerable = descriptor.enumerable || !1, descriptor.configurable = !0, 
                "value" in descriptor && (descriptor.writable = !0), Object.defineProperty(target, descriptor.key, descriptor);
            }
        }(Constructor.prototype, protoProps), Constructor;
    }
    var I18n = function() {
        function I18n(translations, options) {
            this._translations = {}, this._locale = null == options ? void 0 : options.defaultLocale, 
            translations && (this._translations = translations, (null == options ? void 0 : options.defaultLocale) || (this._locale = Object.keys(translations)[0])), 
            this.options = options || {};
        }
        var _proto = I18n.prototype;
        return _proto.t = function(scope, variables) {
            var value, translation = this._translations[this._locale];
            if (!translation) return "function" == typeof this.options.customizeMissing ? this.options.customizeMissing(scope, variables) : '[missing "' + this._locale + '" locale]';
            if ("string" == typeof scope && scope.includes(".") && !(scope in translation)) {
                var _this__findChainValue = this._findChainValue(translation, scope), v = _this__findChainValue[0];
                if (_this__findChainValue[1]) value = v; else {
                    if (void 0 === (null == variables ? void 0 : variables.defaultvalue)) return "function" == typeof this.options.customizeMissing ? this.options.customizeMissing(scope, variables) : '[missing "' + this._locale + "." + scope + '" translation]';
                    value = variables.defaultvalue;
                }
            } else if (scope in translation) value = translation[scope]; else {
                if (void 0 === (null == variables ? void 0 : variables.defaultvalue)) return "function" == typeof this.options.customizeMissing ? this.options.customizeMissing(scope, variables) : '[missing "' + this._locale + "." + scope + '" translation]';
                value = variables.defaultvalue;
            }
            return "string" == typeof value && variables && /\{\{\s*(\w+)\s*\}\}/.test(value) ? value.replace(/\{\{\s*(\w+)\s*\}\}/g, function(_, varName) {
                return String(variables[varName]) || "";
            }) : value;
        }, _proto.appendTranslations = function(translations) {
            this._translations = deepmerge(this._translations, translations, {
                clone: !1
            }), this._locale || (this._locale = Object.keys(this._translations)[0]), this._onChange();
        }, _proto.switchLocale = function(locale) {
            locale in this._translations && (this._locale = locale, this._onChange());
        }, _proto.switchTranslation = function(locale) {
            this.switchLocale(locale);
        }, _proto.getCurrentLocale = function() {
            return this._locale;
        }, _proto.getCurrentTranslation = function() {
            return this._translations[this._locale];
        }, _proto.getVersion = function() {
            return "2.0.0";
        }, _proto._onChange = function() {
            return null == this.options.onChange ? void 0 : this.options.onChange.call(this.options, this);
        }, _proto._findChainValue = function(obj, path) {
            for (var paths = path.split("."), translation = obj, i = 0; i < paths.length; i++) {
                if (!(paths[i] in translation)) return [ void 0, !1 ];
                if (translation = translation[paths[i]], i + 1 === paths.length) return [ translation[paths[i]], !0 ];
            }
            return [ void 0, !1 ];
        }, _create_class(I18n, [ {
            key: "locale",
            get: function() {
                return this._locale;
            }
        }, {
            key: "translations",
            get: function() {
                return this._translations;
            }
        } ]), I18n;
    }();
    return I18n.VERSION = "2.0.0", I18n;
});
